File: client/computerpets_client/gait.py
"""Shared living-desk numbers — same as web ``gait.ts`` and Electron ``gait.js``."""

from __future__ import annotations

ACCEL_S = 0.4
DECEL_DIST = 56.0
TURN_S = 0.23
TURN_SNAKE_S = 0.35
TURN_SLOW_S = 0.3
OVERSHOOT = 7.0
OVERSHOOT_SLOW = 4.0
SETTLE_S = 0.42
POSE_HOLD_S = 0.28
BREATHE_IDLE = 0.028
BREATHE_SLEEP = 0.04
SWAY_PX = 3.5
HIGH_WALK = 120.0


def smoothstep(t: float) -> float:
    x = max(0.0, min(1.0, t))
    return x * x * (3.0 - 2.0 * x)


def walk_speed(remaining: float, age: float, base: float) -> float:
    accel = min(1.0, age / ACCEL_S)
    decel = smoothstep(remaining / DECEL_DIST) if remaining < DECEL_DIST else 1.0
    return base * max(0.3, accel * decel)


def is_low_walk(walk: float, hop: float | None = None) -> bool:
    if hop is not None:
        return hop <= 6 or walk < 40
    return walk < 40


def is_high_walk(walk: float) -> bool:
    return walk >= HIGH_WALK


def turn_hold_s(*, crawl: bool, walk: float, hop: float | None = None) -> float:
    if crawl:
        return TURN_SNAKE_S
    if is_low_walk(walk, hop):
        return TURN_SLOW_S
    return TURN_S


def overshoot_px(*, crawl: bool, walk: float, hop: float | None = None) -> float:
    if crawl or is_low_walk(walk, hop):
        return OVERSHOOT_SLOW
    return OVERSHOOT


def facing_after(current: int, x: float, target: float, elapsed: float, hold_s: float) -> int:
    desired = 1 if target >= x else -1
    if desired == current:
        return current
    if elapsed < hold_s:
        return current
    return desired


def settle_offset(settle: float, direction: int, px: float) -> float:
    t = 1.0 - max(0.0, min(1.0, settle))
    return direction * px * 4.0 * t * (1.0 - t)


def wander_pause_s() -> float:
    import random

    return 0.2 + random.random() * 0.2

File: web/src/lib/pets/gait.ts
import { isSnake } from "./shed";

/** Shared living-desk numbers — ported to Electron `gait.js` and PyQt `gait.py`. */
export const ACCEL_S = 0.4;
export const DECEL_DIST = 56;
export const TURN_S = 0.23;
export const TURN_SNAKE_S = 0.35;
export const TURN_SLOW_S = 0.3;
export const OVERSHOOT = 7;
export const OVERSHOOT_SLOW = 4;
export const SETTLE_S = 0.42;
export const LAND_DECAY = 1 / 0.45;
export const POSE_HOLD_S = 0.28;
export const BREATHE_IDLE = 0.028;
export const BREATHE_SLEEP = 0.04;
export const SWAY_PX = 3.5;
export const WALK_HOP_PX = 4;
export const PERCH_STEP_PX = 7;
export const HIGH_HOP = 20;
export const HIGH_WALK = 120;
export const STEP_S = 0.22;
export const STEP_S_QUICK = 0.16;

export function smoothstep(t: number) {
  const x = Math.max(0, Math.min(1, t));
  return x * x * (3 - 2 * x);
}

export function walkSpeed(remaining: number, age: number, base: number) {
  const accel = Math.min(1, age / ACCEL_S);
  const decel = remaining < DECEL_DIST ? smoothstep(remaining / DECEL_DIST) : 1;
  return base * Math.max(0.3, accel * decel);
}

export function isCrawlKey(key: string | undefined | null) {
  return !!key && isSnake(key);
}

export function isLowWalk(hop: number, walk: number) {
  return hop <= 6 || walk < 40;
}

export function isHighWalk(walk: number) {
  return walk >= HIGH_WALK;
}

export function turnHoldS(opts: { crawl?: boolean; hop?: number; walk?: number }) {
  if (opts.crawl) return TURN_SNAKE_S;
  if (isLowWalk(opts.hop ?? 20, opts.walk ?? 98)) return TURN_SLOW_S;
  return TURN_S;
}

export function overshootPx(opts: { crawl?: boolean; hop?: number; walk?: number }) {
  if (opts.crawl || isLowWalk(opts.hop ?? 20, opts.walk ?? 98)) return OVERSHOOT_SLOW;
  return OVERSHOOT;
}

export function facingAfter(current: 1 | -1, x: number, target: number, elapsed: number, holdS: number): 1 | -1 {
  const desired: 1 | -1 = target >= x ? 1 : -1;
  if (desired === current) return current;
  if (elapsed < holdS) return current;
  return desired;
}

export function settleOffset(settle: number, dir: 1 | -1, px: number) {
  const t = 1 - Math.max(0, Math.min(1, settle));
  return dir * px * 4 * t * (1 - t);
}

export function wanderPauseS() {
  return 0.2 + Math.random() * 0.2;
}

const SPRITE = 176;

/** Walk off the nearest edge. Same sit on desk, overlay, and blotter. */
export function leaveTarget(x: number, width: number, sprite = SPRITE) {
  return x + sprite / 2 < width / 2 ? -sprite - 24 : width + 12;
}

/** Come back from off-stage. Same sit as the desk enter. */
export function enterSpawn(width: number, sprite = SPRITE, pad = 20, left?: boolean) {
  const max = width - sprite - pad;
  const fromLeft = left ?? Math.random() < 0.5;
  return fromLeft ? -sprite : max + sprite;
}

export function enterSit(width: number, sprite = SPRITE, pad = 20, rand = Math.random()) {
  const max = width - sprite - pad;
  const sit = 80 + rand * Math.max(40, max - 80);
  return Math.max(pad, Math.min(max, sit));
}
